File: tool/verify.dart
// tool/verify.dart
//
// The single source of truth for `make verify` (AGENTS.md §0). Runs the full
// gate, in strict order, failing non-zero on the FIRST failure. This is pure
// Dart so it runs identically on Windows, macOS, Linux, and CI — `make verify`
// and the Windows-friendly `dart run tool/verify.dart` execute the exact same
// sequence.
//
// Order (each stage must pass before the next runs):
//   1. dart format --set-exit-if-changed .
//   2. dart analyze --fatal-infos --fatal-warnings
//   3. grep gates            (tool/grep_gates.dart)
//   4. skill-links           (tool/check_skill_links.dart)
//   5. schema fence is fresh (tool/gen_schema.dart --check)
//   6. doc-honesty           (tool/doc_honesty.dart)
//   7. flutter test
//
// A passing run here is the ONLY definition of "done".

import 'dart:io';

Future<void> main() async {
  final stages = <_Stage>[
    _Stage('format', 'dart', ['format', '--set-exit-if-changed', '.']),
    _Stage('analyze', 'dart', ['analyze', '--fatal-infos', '--fatal-warnings']),
    _Stage('grep-gates', 'dart', ['run', 'tool/grep_gates.dart']),
    _Stage('skill-links', 'dart', ['run', 'tool/check_skill_links.dart']),
    _Stage('schema-fresh', 'dart', ['run', 'tool/gen_schema.dart', '--check']),
    _Stage('doc-honesty', 'dart', ['run', 'tool/doc_honesty.dart']),
    _Stage('test', 'flutter', ['test']),
  ];

  final sw = Stopwatch()..start();
  for (final stage in stages) {
    stdout.writeln('\n=== verify: ${stage.name} ===');
    final code = await stage.run();
    if (code != 0) {
      stdout.writeln('\n✗ verify FAILED at stage "${stage.name}" '
          '(exit $code) after ${sw.elapsed.inSeconds}s.');
      exit(code);
    }
  }
  stdout.writeln('\n✓ verify PASSED — all stages green in ${sw.elapsed.inSeconds}s.');
  exit(0);
}

class _Stage {
  _Stage(this.name, this.executable, this.args);
  final String name;
  final String executable;
  final List<String> args;

  Future<int> run() async {
    try {
      final proc = await Process.start(
        executable,
        args,
        mode: ProcessStartMode.inheritStdio,
        runInShell: true,
      );
      return proc.exitCode;
    } on ProcessException catch (e) {
      stderr.writeln('verify: could not run "$executable ${args.join(' ')}": '
          '${e.message}');
      stderr.writeln('Is the Flutter/Dart SDK on your PATH?');
      return 127;
    }
  }
}

File: backend/tool/verify.ts
#!/usr/bin/env -S deno run --allow-all
/**
 * tool/verify.ts — TheLIST backend verify pipeline.
 *
 * Runs all stages in strict fail-fast order. This is the single trustworthy
 * signal: a task is not done until this exits 0.
 *
 * Stages:
 *   1. fmt       — deno fmt --check .
 *   2. lint      — deno lint
 *   3. grep-gates — tool/grep_gates.ts
 *   4. schema-fresh — tool/gen_schema.ts --check
 *   5. doc-honesty — tool/doc_honesty.ts
 *   5b. doc-coverage — tool/doc_coverage.ts
 *   6. test      — deno test --allow-all supabase/functions/
 */

import { join } from "https://deno.land/std@0.224.0/path/mod.ts";

const ROOT = new URL("../", import.meta.url).pathname;

interface Stage {
  name: string;
  cmd: string[];
}

const stages: Stage[] = [
  { name: "fmt", cmd: ["deno", "fmt", "--check", "."] },
  { name: "lint", cmd: ["deno", "lint"] },
  { name: "grep-gates", cmd: ["deno", "run", "--allow-all", join(ROOT, "tool/grep_gates.ts")] },
  { name: "schema-fresh", cmd: ["deno", "run", "--allow-all", join(ROOT, "tool/gen_schema.ts"), "--check"] },
  { name: "doc-honesty", cmd: ["deno", "run", "--allow-all", join(ROOT, "tool/doc_honesty.ts")] },
  { name: "doc-coverage", cmd: ["deno", "run", "--allow-all", join(ROOT, "tool/doc_coverage.ts")] },
  { name: "test", cmd: ["deno", "test", "--allow-all", join(ROOT, "supabase/functions/")] },
];

let failed = false;

for (const stage of stages) {
  const label = `[${stage.name}]`;
  console.log(`\n${label} running: ${stage.cmd.join(" ")}`);

  const proc = new Deno.Command(stage.cmd[0], {
    args: stage.cmd.slice(1),
    cwd: ROOT,
    stdout: "inherit",
    stderr: "inherit",
  });

  const { code } = await proc.output();

  if (code !== 0) {
    console.error(`\n✗ ${label} FAILED (exit ${code}). Stopping.`);
    failed = true;
    break;
  }

  console.log(`✓ ${label} passed`);
}

if (!failed) {
  console.log("\n✓ All stages green. Done.\n");
} else {
  Deno.exit(1);
}
